internals/secrethub/tree.go

```diff
@@ -81,11 +99,11 @@ func printDirContentsRecursively(dir *api.Dir, prefix string, w io.Writer) {
 		name := colorizeByStatus(sub.Status, sub.Name)
 
 		if i == total-1 {
-			fmt.Fprintf(w, "%s└── %s/\n", prefix, name)
-			printDirContentsRecursively(sub, prefix+"    ", w)
+			fmt.Fprintf(w, format[0], prefix, name)
```

Comment:
I'm struggling a bit to read from the `format` variable what is going to be printed in this case (it's defined in another function, `printTree`). How would you feel about defining the format in this function? One way I can imagine is the following:

```Golang
if cmd.noIndentation {
    fmt.Print(w, name)
} else if i == total-1 {
    fmt.Fprintf(w, "%s└── %s/\n", prefix, name)
    prefix += "    "
} else {
    fmt.Fprintf(w, "%s├── %s/\n", prefix, name)
    prefix += "│   "
}
printDirContentsRecursively(sub, prefix, w)
```

And maybe we can put this in a separate function so that we can re-use the same thing for directories and secrets? It looks like they both use the same format in the same cases.

Reply:
I believe the reasoning for having `format` where it currently is starts from the fact that, in this way, we can avoid having to do the check for `noIndent` every time the recursive function calls itself. By having it declared before the first call of the function, we can decide beforehand which prefixes are supposed to be used. (therefore we only create 1 decision point based on `noIndent`, instead of O(n)) The alternative to this would probably be having the prefixes as global variables, but I think this is not a very good practice. Maybe @edif2008 can confirm/chime in? Let us know what you think, @SimonBarendse.

Reply:
You are right that the current code is slightly faster than the proposed changes. However, as far as I can see they both have the same runtime complexity (constant within the function + same recursive call), meaning they grow at the same pace.

What's more, the function is always called with a relatively low `n`. Therefore the function is always fast and the runtime complexity isn't too important. No human would be able to notice the difference.
Therefore, I'd pick readability over speed in this trade-off.

To get a feeling for the speed difference between the two, I've created the following benchmark:
```Golang
package main

import (
	"testing"
)

func BenchmarkWithCondition(b *testing.B) {
	// run the Fib function b.N times
	for n := 0; n < b.N; n++ {
		withCondition(true)
	}
}

func BenchmarkWithLookup(b *testing.B) {
	// run the Fib function b.N times
	for n := 0; n < b.N; n++ {
		withLookup([1]string{"foo"})
	}
}

func withCondition(condition bool) string {
	if condition {
		return "foo"
	}
	return "bar"
}

func withLookup(result [1]string) string {
	return result[0]
}
```

These were my results:
```
BenchmarkWithCondition-8   	1000000000	         0.248 ns/op
BenchmarkWithLookup-8      	1000000000	         0.245 ns/op
```

This means there's `0.000000000003 seconds` difference between evaluating one secret/directory. In a tree with a thousand secrets and directories (this is already a huge tree), this would make a difference of `0.000000003 seconds`.

Reply:
Indeed! Thank you for the clarification. This should now be fixed.

Reply:
The reason we've started with this change is to improve readability by making clear which "format" is used when it is printed.

The last change, switching to `format1`, `format2`, `format3` variables, goes back to having to lookup or guess what this format will be. Can we either use the format directly where it is printed (no variable) or change the name of the constants to something by which you can recognize it (I think this is really hard to get clear with naming, which is why I'm leaning towards defining the "format" strings directly in the print function call.

If you want to prevent duplication of the format strings between the secret and directory loop, can we extract that shared functionality between secrets and directories to a function to achieve that?

Reply:
Will push the version with the formats directly in the print commands first. Meanwhile, I will look up ways to extract the shared functionality

Reply:
Sounds good :smile:

If extracting the shared functionality doesn't make the code more elegant, I'm also okay with having a small duplication.

Reply:
I looked into whether we can extract the shared functionality in an elegant way. Unfortunately, there are a couple of key differences that will not enable an elegant extraction: the difference between a secret and a directory, the fact that we add an extra `/` for the directory. This leaves us to only the checks which, although they look similar, if we extract them, we add a couple of more checks that will not make the code nice in the end (especially the readability). Therefore, we will leave it as it is, since the complexity is small and easy to understand and follow.